### ISSUE TITLE
`filter_map_bool_then` triggers in some cases when variable is borrowed in predicate and transformation.

### ISSUE THREAD

**Issue by dsemi** [rust-lang/rust-clippy#11617]
### Summary

In cases where a variable needs to be borrowed by both the predicate function in `filter` and the transformation function in `map`, and one of those borrows is mutable, the `filter_map` -> `filter` + `map` transformation is invalid.

### Lint Name

filter_map_bool_then

### Reproducer

I tried this code:

```rust
fn test() {
    let mut x: Vec<usize> = vec![0; 10];
    let _ = (0..x.len()).filter_map(|i| {
        (x[i] != i).then(|| {
            x[i] = i;
            i
        })
    });
}
```

I expect this to not trigger a lint warning because the transformation is invalid. The transformed code looks like:

```rust
fn test() {
    let mut x: Vec<usize> = vec![0; 10];
    let _ = (0..x.len()).filter(|&i| x[i] != i).map(|i| {
        x[i] = i;
        i
    });
}
```

This fails to compile because `x` is borrowed in `filter` and mutably borrowed in `map`:

```
error[E0502]: cannot borrow `x` as mutable because it is also borrowed as immutable
  --> src/main.rs:42:53
   |
42 |     let _ = (0..x.len()).filter(|&i| x[i] != i).map(|i| {
   |                                 ---- -          --- ^^^ mutable borrow occurs here
   |                                 |    |          |
   |                                 |    |          immutable borrow later used by call
   |                                 |    first borrow occurs due to use of `x` in closure
   |                                 immutable borrow occurs here
43 |         x[i] = i;
   |         - second borrow occurs due to use of `x` in closure

For more information about this error, try `rustc --explain E0502`.
```

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: aarch64-apple-darwin
release: 1.73.0
LLVM version: 17.0.2
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
